## Use Time series interpolation method when the index is In Date format

In [70]:
import pandas as pd

In [71]:
import numpy as np

In [72]:
data={"price":[103,405,606,np.nan,540,657,np.nan,np.nan,412]}
date=pd.date_range('2025-10-03',periods=9,freq='d')
df=pd.DataFrame(data,index=date)

print("Before Interpolation")
print(df)




Before Interpolation
            price
2025-10-03  103.0
2025-10-04  405.0
2025-10-05  606.0
2025-10-06    NaN
2025-10-07  540.0
2025-10-08  657.0
2025-10-09    NaN
2025-10-10    NaN
2025-10-11  412.0


In [73]:
print("After Interpolation")
new_df=df[['price']].interpolate(method="time",axis=0,inplace=False)
print(new_df)

After Interpolation
                 price
2025-10-03  103.000000
2025-10-04  405.000000
2025-10-05  606.000000
2025-10-06  573.000000
2025-10-07  540.000000
2025-10-08  657.000000
2025-10-09  575.333333
2025-10-10  493.666667
2025-10-11  412.000000


# 🧮 Linear / Time Interpolation Calculation (Manual)

## 📅 Given Dataset

| Date | Price |
|-------|--------|
| 2025-10-03 | 103.0 |
| 2025-10-04 | 405.0 |
| 2025-10-05 | 606.0 |
| 2025-10-06 | NaN ❌ |
| 2025-10-07 | 540.0 |
| 2025-10-08 | 657.0 |
| 2025-10-09 | NaN ❌ |
| 2025-10-10 | NaN ❌ |
| 2025-10-11 | 412.0 |

---

## 🧠 Interpolation Formula

**Formula:**
Y = Y1 + ((X - X1) / (X2 - X1)) * (Y2 - Y1)


Where:  
- Y1, Y2 → Known data values  
- X1, X2 → Corresponding time/index values  
- X → Missing time point  
- Y → Interpolated value  

> Since all dates are equally spaced (1-day apart),  
> both `linear` and `time` interpolation give the **same result**.

---

## 🔹 Step 1: Interpolate for 2025-10-06

**Known values:**
- X1 = 2025-10-05, Y1 = 606  
- X2 = 2025-10-07, Y2 = 540  
- X = 2025-10-06  

**Calculation:**
Y = 606 + ((1) / (2)) * (540 - 606)
Y = 606 - 33
Y = 573.0



✅ **Interpolated (2025-10-06) = 573.0**

---

## 🔹 Step 2: Interpolate for 2025-10-09 and 2025-10-10

**Known values:**
- X1 = 2025-10-08, Y1 = 657  
- X2 = 2025-10-11, Y2 = 412  

**Total difference:**
Y2 - Y1 = 412 - 657 = -245



**Step difference (3 intervals):**
-245 / 3 = -81.67



| Date | Formula | Value |
|-------|----------|--------|
| 2025-10-08 | Given | 657.00 |
| 2025-10-09 | 657 - 81.67 | 575.33 |
| 2025-10-10 | 575.33 - 81.67 | 493.66 |
| 2025-10-11 | Given | 412.00 |

✅ **Interpolated (2025-10-09) = 575.33**  
✅ **Interpolated (2025-10-10) = 493.67**

---

## ✅ Final Interpolated Table

| Date | Price |
|-------|--------|
| 2025-10-03 | 103.0 |
| 2025-10-04 | 405.0 |
| 2025-10-05 | 606.0 |
| 2025-10-06 | **573.0** |
| 2025-10-07 | 540.0 |
| 2025-10-08 | 657.0 |
| 2025-10-09 | **575.33** |
| 2025-10-10 | **493.67** |
| 2025-10-11 | 412.0 |

---

## 💡 Summary

| Point | Description |
|--------|--------------|
| Formula Used | Y = Y1 + ((X - X1) / (X2 - X1)) * (Y2 - Y1) |
| Interpolation Type | Linear or Time (both same here due to equal 1-day gaps) |
| Interpolated Points | 2025-10-06 → 573.0, 2025-10-09 → 575.33, 2025-10-10 → 493.67 |
| Key Note | When time gaps are equal, `method='linear'` and `method='time'` give identical results. |
